secant/Features/BackupFlow/RecoveryPhraseValidation.swift

```diff
@@ -0,0 +1,111 @@
+//
+
+import Foundation
+import ComposableArchitecture
+
+typealias RecoveryPhraseValidationStore = Store<RecoveryPhraseValidationState, RecoveryPhraseValidationAction>
+typealias RecoveryPhraseValidationViewStore = ViewStore<RecoveryPhraseValidationState, RecoveryPhraseValidationAction>
+
+struct RecoveryPhraseEnvironment {
+    var mainQueue: AnySchedulerOf<DispatchQueue>
+    var newPhrase: () -> Effect<RecoveryPhrase, AppError>
+}
+
+struct RecoveryPhraseValidationState: Equatable {
+    static let wordGroupSize = 6
+    static let phraseChunks = 4
+
+    var step: RecoveryPhraseValidationStep
+
+    static func pickWordsFromMissingIndices(indices: [Int], phrase: RecoveryPhrase) -> [PhraseChip.Kind] {
+        precondition((indices.count - 1) * wordGroupSize <= phrase.words.count)
+        var words: [PhraseChip.Kind] = []
+        indices.enumerated().forEach({ index, position in
+            let realIndex = (index * wordGroupSize) + position
+            words.append(.unassigned(word: phrase.words[realIndex]))
+        })
+        return words
+    }
+
+    static func randomIndices() -> [Int] {
+        Array(repeating: Int.random(in: 0...wordGroupSize - 1), count: phraseChunks)
+    }
+
+    static func firstStep(phrase: RecoveryPhrase) -> RecoveryPhraseValidationStep {
+        let missingIndices = Self.randomIndices()
+        let missingWordChipKind = Self.pickWordsFromMissingIndices(indices: missingIndices, phrase: phrase).shuffled()
+        return .initial(phrase: phrase, missingIndices: missingIndices, missingWordsChips: missingWordChipKind)
+    }
+
+    init(phrase: RecoveryPhrase) {
+        let missingIndices = Self.randomIndices()
+        let missingWordChipKind = Self.pickWordsFromMissingIndices(indices: missingIndices, phrase: phrase).shuffled()
+        self.step = .initial(phrase: phrase, missingIndices: missingIndices, missingWordsChips: missingWordChipKind)
+    }
+    /**
+    reset the state to the initial step
+    */
+    mutating func reset() {
+        switch self.step {
+        case let .initial(phrase, _, _):
+            self.step = Self.firstStep(phrase: phrase)
+
+        case .incomplete(let phrase, _, _, _):
+            self.step = Self.firstStep(phrase: phrase)
+
+        case .complete(let phrase, _, _, _):
+            self.step = Self.firstStep(phrase: phrase)
+
+        case .valid(let phrase, _, _, _):
+            self.step = Self.firstStep(phrase: phrase)
+            
+        case .invalid(let phrase, _, _, _):
+            self.step = Self.firstStep(phrase: phrase)
+        }
+    }
+    /**
+    call this when the user drops an unassigned word chip into a group/
+    */
+    mutating func apply(chip: PhraseChip.Kind, into group: Int) {
+        self.step = RecoveryPhraseValidationStep.given(step: self.step, apply: chip, into: group)
+    }
+}
+
+enum RecoveryPhraseValidationAction: Equatable {
+    case reset
+    case drag(wordChip: PhraseChip.Kind, intoGroup: Int)
+    case validate
```

Comment:
It could be possible that if you move to a struct based value with computed properties, this would end up having only two cases, as the validate succeed and fail might all happen automatically with the computed values.

Reply:
I'm definitely getting there, but I'm having doubts on how to achieve a proper transition to either success of fail screens upon completion. How that end up being will definitely shape this

Reply:
So when you drag that last word into place, if it's all correct, everything is valid and the screen automatically performs a transition?

Reply:
Finally the states are
````Swift
enum Step: Equatable {
        case initial
        case incomplete
        case complete
    }
````